### PR TITLE
Create feature toggles

### PR DESCRIPTION
**Pull request by Luiz-FS**
**Feature/Bug description:** Create feature management service

**Solution:** Create a backend template to store and manage system features, and a frontend service that allows you to check whether features are enabled or not.

**TODO/FIXME:** n/a

### CODE REVIEW

File: frontend/utils/featureToggleService.js [hunk 1]
```diff
@@ -0,0 +1,43 @@
...
+
+
+        /**
+         * This function checks whether the feature passed by parameter 
+         * is enabled for the logged in user and the device that he uses.
+         * 
+         * @param {String} featureName - Feature name to check if is enabled
+         * @return {Promise} Promise that when it is resolved it returns a boolean indicating whether or not it is enabled.
+         */
+        service.isEnabled = function isEnabled(featureName) {
```
Comment (JuliePessoa) [inline frontend/utils/featureToggleService.js]: what happens if featureName is undefined?

File: backend/handlers/feature_toggle_handler.py [hunk 2]
```diff
@@ -0,0 +1,50 @@
+"""Feature Toogle handler."""
```
Comment (raonismaneoto) [inline backend/handlers/feature_toggle_handler.py]: toggle *

File: backend/handlers/feature_toggle_handler.py [hunk 3]
```diff
@@ -0,0 +1,50 @@
...
+    """
+    
+    features = [feature.make() for feature in feature_list]
+    return features
+
+class FeatureToggleHander(BaseHandler):
+    """Feature toggle hanler."""
+    
+    @json_response
+    def get(self):
```
Comment (raonismaneoto) [inline backend/handlers/feature_toggle_handler.py]: isn't important to decorate this method with login_required?

File: backend/handlers/feature_toggle_handler.py [hunk 4]
```diff
@@ -0,0 +1,50 @@
...
+    
+    @login_required
+    @json_response
+    def put(self, user):
+        """
+        Method for modifying the properties of one or more features.
+        """
+
+        features_body = json.loads(self.request.body)
+        features = Feature.enable_all(features_body)
```
Comment (raonismaneoto) [inline backend/handlers/feature_toggle_handler.py]: why enable_all? What if the client wants to disable a feature? Or just change other thing?

File: backend/handlers/feature_toggle_handler.py [hunk 5]
```diff
@@ -0,0 +1,50 @@
...
+        self.response.write(json.dumps(to_json(features)))
```
Comment (raonismaneoto) [inline backend/handlers/feature_toggle_handler.py]: you could call json.dumps inside of to_json

File: backend/handlers/feature_toggle_handler.py [hunk 6]
```diff
@@ -0,0 +1,50 @@
...
+    Method to generate list of feature models in json format object.
+
+    Params:
+    feature_list -- List of features objects
+    """
+    
+    features = [feature.make() for feature in feature_list]
+    return features
+
+class FeatureToggleHander(BaseHandler):
```
Comment (raonismaneoto) [inline backend/handlers/feature_toggle_handler.py]: Handler*

File: backend/models/feature.py [hunk 7]
```diff
@@ -0,0 +1,104 @@
...
+    """
+
+    name = ndb.StringProperty()
+    enable_mobile = ndb.StringProperty(
+        choices=set(["SUPER_USER", "ADMIN", "ALL", "DISABLED"]))
+    enable_desktop = ndb.StringProperty(
+        choices=set(["SUPER_USER", "ADMIN", "ALL", "DISABLED"]))
+
+    @staticmethod
+    @ndb.transactional(xg=True)
```
Comment (raonismaneoto) [inline backend/models/feature.py]: why transactional in a staticmethod?

File: backend/models/feature.py [hunk 8]
```diff
@@ -0,0 +1,104 @@
...
+        """
+
+        feature = Feature(id=name, name=name, enable_desktop=enable_desktop, enable_mobile=enable_mobile)
+        feature.put()
+        return feature
+    
+    @staticmethod
+    def enable_all(features_list):
+        """
+        Method to enable or disable multiple features.
```
Comment (raonismaneoto) [inline backend/models/feature.py]: if it enables or disables the name is wrong

File: frontend/app.js [hunk 9]
```diff
@@ -505,6 +505,29 @@
...
+        }, function (transition) {
+            const targetStateName = transition.to().name;
+
+            return FeatureToggleService.isEnabled(MapStateToFeatureService.getFeatureByState(targetStateName)).then(function(enabled) {
+                if (enabled) {
+                    return transition;
+                } else {
+                    return transition.router.stateService.target(STATES.ERROR, {
+                        "msg": "Desculpa! Este recurso ainda não está disponível.",
+                        "status": "500"
```
Comment (raonismaneoto) [inline frontend/app.js]: I think it is not a typical 500 error, which means server error, change to a status that tells more about the specific error

File: frontend/utils/featureToggleService.js [hunk 10]
```diff
@@ -0,0 +1,43 @@
...
+            return service.getFeatures(featureName).then(function(response) {
+                const feature = response[0];
```
Comment (raonismaneoto) [inline frontend/utils/featureToggleService.js]: why response[0]?

File: frontend/utils/featureToggleService.js [hunk 11]
```diff
@@ -0,0 +1,43 @@
...
+                const disabledMobile = feature.enable_mobile === 'DISABLED';
+                const disabeldDesktop = feature.enable_desktop === 'DISABLED';
+                
+                if (disabledMobile && disabeldDesktop)
+                    return false;
+
+                return !disabeldDesktop && !Utils.isMobileScreen() || !disabledMobile && Utils.isMobileScreen();
```
Comment (raonismaneoto) [inline frontend/utils/featureToggleService.js]: disabled*

File: frontend/sw.js [hunk 12]
```diff
@@ -7,7 +7,7 @@
     importScripts('app/firebase-config.js');
     importScripts('app/config.js');
     // if the line number of the code below changes, modify the /ecis script.
-    const CACHE_SUFIX = 'master';
+    const CACHE_SUFIX = 'create-feature-toggles';
```
Comment (raonismaneoto) [inline frontend/sw.js]: undo

File: frontend/test/specs/utils/featureToggleServiceSpec.js [hunk 13]
```diff
@@ -0,0 +1,129 @@
...
+                expect(response).toBeFalsy();
+                done();
+            });
+            httpBackend.flush();
+        });
+
+        it('Should be return false with test-other-feature', function(done) {
+            window.screen = {width: 100};
+            featureToggleService.isEnabled(otherFeature.name).then(function(response) {
+                expect(response).toBeFalsy();
```
Comment (raonismaneoto) [inline frontend/test/specs/utils/featureToggleServiceSpec.js]: toEqual(false) I think is better

**Pull request merged**